# Logistic Regression

- This is a supplement material for my lectures on ML. It sheds light on Python implementations of the supervised machine learning algorithms. 
- I assume you know Python syntax and how it works. If you don't, I highly recommend you to take a break and get introduced to the language before going forward with my code. 
- This material is represented as a Jupyter notebook, which you can easily download to reproduce the code and play around with it. 

## 1. Libraries

To build a model, we need 
- `pandas` library to work with panel dataframes
- `sklearn.model_selection.train_test_split` to split the dataset into train and test
- `sklearn.linear_model.LogisticRegression` to build a model

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

## 2. Data Load & Overview

Let's load the dataset and understand it.

In [51]:
# Load the dataset
df = pd.read_csv('https://raw.githubusercontent.com/5x12/ml-cookbook/master/supplements/data/heart.csv')

In [52]:
# Print the first 10 rows of the dataset
df.head(10)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   
5   58    0   0       100   248    0        0      122      0      1.0      1   
6   58    1   0       114   318    0        2      140      0      4.4      0   
7   55    1   0       160   289    0        0      145      1      0.8      1   
8   46    1   0       120   249    0        0      144      0      0.8      2   
9   54    1   0       122   286    0        0      116      1      3.2      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0  
5   0     2       1  
6   3     1       0  
7   1     3       0  
8   0     3       0  
9   2     2       0

This Public Health Dataset represents people's medical records.

- age: age in years
- sex: (1 = male; 0 = female)
- cp: chest pain type
- trestbps: resting blood pressure (in mm Hg on admission to the hospital)
- chol: serum cholestoral in mg/dl
- fbs: (fasting blood sugar > 120 mg/dl) (1 = true; 0 = false)
- restecg: resting electrocardiographic results
- thalach: maximum heart rate achieved
- exang: exercise induced angina (1 = yes; 0 = no)
- oldpeak: ST depression induced by exercise relative to rest
- slope: the slope of the peak exercise ST segment
- ca: number of major vessels (0-3) colored by flourosopy
- thal: 3 = normal; 6 = fixed defect; 7 = reversable defect
- target: 1 or 0. It refers to the presence of heart disease in the patient. It is integer valued 0 = no disease and 1 = disease.

The main aim is to build **a model that predicts a heart disease of a patient** (target column) based on independent variables.

## 3. Variables

Let's split the dataset into X and y, where 
- $X$ is a set of independent variables (all columns but the last one)
- $y$ is a dependent, or target variable (the last column)

In [53]:
# Filter out target column and take all other columns
X = df.iloc[:, :-1]

# Print X
X

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  \
0      52    1   0       125   212    0        1      168      0      1.0   
1      53    1   0       140   203    1        0      155      1      3.1   
2      70    1   0       145   174    0        1      125      1      2.6   
3      61    1   0       148   203    0        1      161      0      0.0   
4      62    0   0       138   294    1        1      106      0      1.9   
...   ...  ...  ..       ...   ...  ...      ...      ...    ...      ...   
1020   59    1   1       140   221    0        1      164      1      0.0   
1021   60    1   0       125   258    0        0      141      1      2.8   
1022   47    1   0       110   275    0        0      118      1      1.0   
1023   50    0   0       110   254    0        0      159      0      0.0   
1024   54    1   0       120   188    0        1      113      0      1.4   

      slope  ca  thal  
0         2   2     3  
1         0   0     3  
2         0   0     3  
3         2   1     3  
4         1   3     2  
...     ...  ..   ...  
1020      2   0     2  
1021      1   1     3  
1022      1   1     2  
1023      2   0     2  
1024      1   1     3  

[1025 rows x 13 columns]

In [54]:
# Select target column
y = df['target']

# Print y
y

0       0
1       0
2       0
3       0
4       0
       ..
1020    1
1021    0
1022    0
1023    1
1024    0
Name: target, Length: 1025, dtype: int64

## 4. Model

In this section we are going to 
- build a logistic regression model, 
- evaluate its accuracy, and 
- make a prediction.

### 4.1. Building the model

Let's recall three simple steps:

- Split the $X$ & $y$ variables into train and test sets
- Initialize the model
- Train the model with the varialbes

In [87]:
# Split variables into train and test
X_train, X_test, y_train, y_test = train_test_split(X, #independent variables
                                                    y, #dependent variable
                                                    random_state = 3
                                                   )

In [90]:
# Initialize the model
clf = LogisticRegression(random_state = 3)

In [91]:
# Train the model
clf.fit(X_train, y_train)

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=3)

### 4.2. Checking models accuracy

After the model has trained with the data, it's essential to understand how precisely it predicts heart disease. For that, we need to check model's accuracy. 

In [92]:
print(f'Accuracy of Logistic Regression classifier on training set: {clf.score(X_train, y_train):.2f}')
print(f'Accuracy of Logistic Regression classifier on test set: {clf.score(X_test, y_test):.2f}')

Accuracy of Logistic Regression classifier on training set: 0.85
Accuracy of Logistic Regression classifier on test set: 0.83


### 4.3. Making a prediction

Now that we know the model is accurate enough, we can predict whether or not a patient is having a heart disease by passing independent varialbes to the model. The method `predict` returns such a prediction - 0 for NO, 1 for YES.

In [93]:
clf.predict([[59, 1, 0, 101, 234, 0, 1, 143, 0, 3.4, 0, 0, 0]])

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/base.py:409: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


array([0])

In the array:
- value 0 means a patient does not have a heart disease, 
- value 1 means a patient does not have a heart disease.

We can also check the probability of a patient having a heart disease. The method `predict_proba` can be used to infer the class probabilities (i.e. the probability that a particular data point falls into the underlying classes).

In [94]:
clf.predict_proba([[59, 1, 0, 101, 234, 0, 1, 143, 0, 3.4, 0, 0, 0]])

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/base.py:409: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


array([[0.55699362, 0.44300638]])

This array means that there is a 
- 55.699362% chance of having NO heart disease, and 
- 44.300638% chance of having a heart disease. 

Hence, the final prediction will be 0 (NO heart disease).

## 5. Hyperparameters Tuning

Hyperparameter Tuning is one of the critical steps in model building. It helps in increasing model's accuracy even further.

### 5.1. Playing around with one hyperparameter

During model initialization, we should accurately select parameters. Previously above we did not choose any parameters, but let's see what happens if we set one of the regularization parameters `C=1`.

In [74]:
# Initialize the model
clf = LogisticRegression(C=50)

In [75]:
# Train the model
clf.fit(X_train, y_train)

/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(C=50)

In [76]:
# Checking model's accuracy
print(f'Accuracy of Logistic Regression classifier on training set: {clf.score(X_train, y_train):.2f}')
print(f'Accuracy of Logistic Regression classifier on test set: {clf.score(X_train, y_train):.2f}')

Accuracy of Logistic Regression classifier on training set: 0.87
Accuracy of Logistic Regression classifier on test set: 0.87


We see an increase of the model's accuracy. We can also write a simple `for` loop to observe different accuracy scores for different values of `C`, and pick value of `C` that gives the highest accuracy. This is called hyperparameter tuning.

In [77]:
for this_C in [0.1, 1, 100]:
    
    #initialize a model
    clf = LogisticRegression(C=this_C)
    
    #fit the model
    clf.fit(X_train, y_train)
    
    #print results
    print(f'Logistic Regression with C = {this_C}')
    print(f'Accuracy of Logistic regression classifier on training set: {clf.score(X_train, y_train):.2f}')
    print(f'Accuracy of Logistic regression classifier on test set: {clf.score(X_test, y_test):.2f}\n')

Logistic Regression with C = 0.1
Accuracy of Logistic regression classifier on training set: 0.84
Accuracy of Logistic regression classifier on test set: 0.88

Logistic Regression with C = 1
Accuracy of Logistic regression classifier on training set: 0.86
Accuracy of Logistic regression classifier on test set: 0.88

Logistic Regression with C = 100
Accuracy of Logistic regression classifier on training set: 0.86
Accuracy of Logistic regression classifier on test set: 0.87



/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/andrewwolf/.pyenv/versions/3.10.7/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/m

### 5.2. All Hyperparameters of Logistic Regression

There are many hyperparameters in Logistic Regression. All are presented below - as you can see, they have values by default. Feel free to play around with them. To understand each hyperparameter, visit https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html.

In [70]:
clf = LogisticRegression(penalty='l2', 
                         dual=False, 
                         tol=0.0001, 
                         C=1.0,             #what we played around with above
                         fit_intercept=True, 
                         intercept_scaling=1, 
                         class_weight=None, 
                         random_state=None, 
                         solver='lbfgs', 
                         max_iter=100, 
                         multi_class='auto', 
                         verbose=0, 
                         warm_start=False, 
                         n_jobs=None, 
                         l1_ratio=None
                        )